Add location and sign to KS test result

iosonofabio opened the pull request:
#### Reference issue
Aims to close #9389

#### What does this implement/fix?
Moves the result of Kolmogorov-Smirnov tests (1 or 2 samples) to a "tuple bunch" with two additional attributes:
- `statistic_location` indicating at what value of the stochastic variable (approximately) the maximal distance in CDF is reached
- `statistic_sign`, which is only nonempty for `alternative="two-sided"`, indicating whether the first distribution (for 2samp)/the empirical distribution (for 1samp) was above or below the second/theoretical distribution at the `statistic_location`. This is a cheap way to determine not only whether the distros are different, but also if they are, in what direction.

#### Additional information
Matt and I quickly discussed part 1, the `statistic_location`. I later realized having part 2 i.e. the sign is equally important for me and perhaps for other folks, and is a cheap addition in terms of performance.

I haven't adapted the part of the docstring that represents doctests yet, happy to do so but perhaps it'd be useful if one of the maintainers took a quick look the the draft PR first so I know if there are any major issues.

Thank you for your time and patience

mdhaber commented:
@tupui @tirthasheshpatel @chrisb83 If you agree that this looks like a good idea, I can review this. There was a lot of user support in the original issue gh-9389, and they seem happy to have it added to the existing function, so I think it's fine to add these attributes rather than create a new function.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -8098,7 +8167,7 @@ def ks_2samp(data1, data2, alternative='two-sided', method='auto'):
 
     prob = np.clip(prob, 0, 1)
-    return KstestResult(d, prob)
+    return KstestResult(d, prob, statistic_location=d_location, statistic_sign=d_sign)
```
```suggestion
    return KstestResult(d, prob, statistic_location=d_location,
                        statistic_sign=d_sign)
```
PEP8: `scipy/stats/_stats_py.py:8170:80: E501 line too long (86 > 79 characters)`

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7549,7 +7549,8 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
 
 
-KstestResult = namedtuple('KstestResult', ('statistic', 'pvalue'))
+KstestResult = _make_tuple_bunch(['statistic', 'pvalue'],
```
```suggestion
KstestResult = _make_tuple_bunch('KstestResult', ['statistic', 'pvalue'],
```
I think this is causing most of the test failures, if not all.

mdhaber commented:
OK, looks like there's just one real failure now:
https://github.com/scipy/scipy/blob/b5a805248c5c3b13e8d1b6237209854b6d0b4359/scipy/stats/tests/test_mstats_basic.py#L1877-L1882

This `Ks_2sampResult`:
https://github.com/scipy/scipy/blob/2e5883ef7af4f5ed4a5b80a1759a45e43163bf3f/scipy/stats/_stats_py.py#L7489

is no longer used anywhere, so I'd suggest just deleting the definition of it and deleting the test. That looks like it will fix the remaining failures.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7705,18 +7724,32 @@ def ks_1samp(x, cdf, args=(), alternative='two-sided', method='auto'):
+    if Dplus > Dminus:
+        D = Dplus
+        d_location = dplus_location
+        d_sign = '+'
```
What do you think about indicating the sign with `1` and `-1` (number) instead of a string? 
Also, even though it's redundant with the specified `alternative`, I'd suggest including it in the result for the one-sided alternative.

iosonofabio replied:
Sure

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -8054,11 +8099,35 @@ def ks_2samp(data1, data2, alternative='two-sided', method='auto'):
+        d = maxS
+        d_location = loc_maxS
+        d_sign = '+'
+    else:
+        if minS > maxS:
+            d = minS
+            d_location = loc_minS
+            d_sign = '-'
+        else:
+            d = maxS
+            d_location = loc_maxS
+            d_sign = '+'
+
```
```suggestion
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -8054,11 +8099,35 @@ def ks_2samp(data1, data2, alternative='two-sided', method='auto'):
+    minS = np.clip(-cddiffs[argminS], 0, 1)
+    maxS = cddiffs[argmaxS]
+
+    if alternative == 'less':
```
```suggestion
    if alternative == 'less' or minS > maxS:
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -8054,11 +8099,35 @@ def ks_2samp(data1, data2, alternative='two-sided', method='auto'):
+        d = minS
+        d_location = loc_minS
+        d_sign = '-'
+    elif alternative == 'greater':
```
```suggestion
    elif alternative == 'greater' or minS <= maxS:
```
Could just be `else`.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7617,11 +7624,23 @@ def ks_1samp(x, cdf, args=(), alternative='two-sided', method='auto'):
+        pvalue : float
+            One-tailed or two-tailed p-value.
+        statistic_location: float
+            Value of the stochastic variable closest to the KS statistic. By
+            definition, the distance between the theoretical and sampled
+            distributions at this value is largest.
```
"closest to"? 
The KS statistics is the difference between the theoretical and empirical CDFs _at_ this location, no?

iosonofabio replied:
It's mostly wording to ward off people who are worried about aliasing. Happy to rephrase more directly.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7617,11 +7624,23 @@ def ks_1samp(x, cdf, args=(), alternative='two-sided', method='auto'):
+            definition, the distance between the theoretical and sampled
+            distributions at this value is largest.
+        statistic_sign: str
+            For alternative == 'two-sided', this attribute is set to '-' if the
```
I will probably have a suggested wording for this, too. I don't want to write a concrete suggestion yet, but I'm thinking we'll want to say something like `+1` if the statistic is the maximum _positive_ difference between the ____ CDF and the ____ CDF (can't remember which is theoretical vs empirical) and `-1` if the statistic is the maximum _negative_ difference...

tupui commented:
I am +0.5. I personally would go with a separate function. We don't really have a lot of extra info in other tests. Instead we could have a general function to get the distance between 2 distributions (our API is slim there) and it would make perfect sense to have the location there.

iosonofabio commented on scipy/stats/_stats_py.py:
```diff
@@ -7705,18 +7724,32 @@ def ks_1samp(x, cdf, args=(), alternative='two-sided', method='auto'):
     x = np.sort(x)
     cdfvals = cdf(x, *args)
 
+    # TODO: add logic to find the argmax
```
Hehe let me take this out

mdhaber commented:
One part of my hesitation with adding a separate function is that if we were to add a function for the statistic used in each of our hypothesis tests, we'd double the number of functions in `stats`. I doubt we want to head in that direction, and I don't think the KS statistic is so special that is it deserves a function while the rest do not.

Also, it makes me think of Halton/Hammersley - those are actually different, yet we're talking about combining them. Here, we'd duplicate the function that computes the statistic, and the only difference between the functions would be that one version has the location of the maximum difference and the other has a p-value?

tupui commented:
I agree that we don't want to duplicate the API with a location function for all tests. 

Here it's just that for this specific test we do something very generic and useful otherwise which is to compute the distance between CDFs.

I am fine doing both as well. Adding now an output for the location and maybe latter have a separate distance function.

iosonofabio commented:
We could make an optional input argument called `compute_Pvalue=True` that skips the last part and is used just to compute the max distance and its location. Would that be a useful compromise in your opinion or just a mess? :smirk:

mdhaber commented:
Mmm I'm more hesitant to add inputs to functions than outputs, and even more hesitant to add an input that instructs the function to add an output : )

It sounds like we have an agreement here - what @tupui said.
> Adding now an output for the location and maybe latter have a separate [CDF] distance function.

mdhaber commented:
So to sum up @iosonofabio we can continue with this PR. Please consider the adjustments recommended above and I can take a second pass. I'd also suggest adding a unit test in `test_stats.py`. a good test would be to confirm that the statistic and p-value result object can still be unpacked and match the attributes, and also that the new attributes are present and appropriate (if possible).

iosonofabio commented:
No problem, I'll get it done as soon as I get back to Oz (haven't slept in a while now...)

Thank you!

mdhaber commented:
I saw you posted some new commits - did you want this to be reviewed again?

iosonofabio commented:
Thanks @mdhaber - I was just overwhelmed in the last few days.

I have now added decent test functions, improved docstrings a little, and changed a piece of logic that you suggested and I believe was incorrect (please double check if you have time!).

If it passes CI, it is ready for you to take a look, thank you! Otherwise I'll just fix the errors first.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7617,11 +7628,22 @@ def ks_1samp(x, cdf, args=(), alternative='two-sided', method='auto'):
+            definition, the distance between the theoretical and sampled
+            distributions is largest at this location.
+        statistic_sign: str
+            1 if the KS statistic is the maximal positive difference between
+            empirical and theoretical distribution, -1 if the KS statistic is
+            the negative difference.
```
```suggestion
            1 if the KS statistic is the maximal positive difference between
            empirical and hypothesized distributions (D+), -1 if the KS
            statistic is the negative difference (D-).
```
Because "D" is ultimately either D+ or D-, do you think we should alter the description above to eliminate "D"?

> KS test statistic, either D+ or D- (depending on the value of 'alternative')

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7920,10 +7956,20 @@ def ks_2samp(data1, data2, alternative='two-sided', method='auto'):
+            definition, the distance between the two sampled distributions at
+            this value is largest.
+        statistic_sign: int
+            +1 if the first distribution is above the second at the KS
+            statistic location, otherwise -1.
```
```suggestion
            +1 if the empirical distribution function of `data1` exceeds
            the empircal distribution function of `data2` at 
            `statistic_location`, otherwise -1.
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7920,10 +7956,20 @@ def ks_2samp(data1, data2, alternative='two-sided', method='auto'):
+        pvalue : float
+            One-tailed or two-tailed p-value.
+        statistic_location: float
+            Value of the stochastic variable closest to the KS statistic. By
+            definition, the distance between the two sampled distributions at
+            this value is largest.
```
```suggestion
            Value from `data1` or `data2` corresponding with the KS statistic;
            i.e., the distance between the empirical distribution functions is
            greatest at this observation.
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7617,11 +7628,22 @@ def ks_1samp(x, cdf, args=(), alternative='two-sided', method='auto'):
+        statistic_location: float
+            Value of the stochastic variable at the KS statistic. By
+            definition, the distance between the theoretical and sampled
+            distributions is largest at this location.
```
Also, what do you think about "observation" instead of "location"? This is reinforce the idea that this is an element from the sample.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3982,12 +3995,22 @@ def testTwoVsThree(self):
-        self._testOne(data1m, data2, 'two-sided', 2.0 / 3, 0.6)
-        self._testOne(data1m, data2, 'greater', 2.0 / 3, 0.3)
-        self._testOne(data1m, data2, 'less', 0, 1.0)
+        self._testOne(data1p, data2, 'two-sided', 1.0 / 3, 1.0,
+                      attrs={'statistic_location': 2.01, 'statistic_sign': 1})
+        self._testOne(data1p, data2, 'greater', 1.0 / 3, 0.7,
+                      attrs={'statistic_location': 2.01, 'statistic_sign': 1})
+        self._testOne(data1p, data2, 'less', 1.0 / 3, 0.7,
+                      attrs={'statistic_location': 1, 'statistic_sign': -1})
+        self._testOne(data1m, data2, 'two-sided', 2.0 / 3, 0.6,
+                      attrs={'statistic_location': 1.99, 'statistic_sign': 1})
+        self._testOne(data1m, data2, 'greater', 2.0 / 3, 0.3,
+                      attrs={'statistic_location': 1.99, 'statistic_sign': 1})
+        # The distributions never cross, so their max distance is zero which is
+        # true beyond both edges of the union (in practice, 0.99 and 3.0 are
+        # both correct for "statistic_location"). Therefore, we do not check
+        # for statistic_location to avoid future false positive test failures.
+        self._testOne(data1m, data2, 'less', 0, 1.0,
+                      attrs={'statistic_sign': -1})
```
I'm not sure this is the strongest test. If you already thought through each one and reasoned out the answer (rather than just relying on the output of `ks_2samp` to tell you what the reference values should be), we can leave it.

Personally, though, I'd prefer to construct a test that makes it obvious what the answers should be. For instance, perhaps we would construct two samples that correspond very closely except at one observation. 
```python3
x = np.arange(10, dtype=np.float64)
y = x.copy()
y[6] = 6.9
stats.ks_2samp(x, y)
```

Something similar could be done for `ks_1samp`.

iosonofabio replied:
done now

mdhaber reviewed (commented):
Thanks @iosonofabio. The implementation looks perfect. My only comments are about the documentation and the tests.

iosonofabio commented on scipy/stats/_stats_py.py:
```diff
@@ -8177,10 +8238,18 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', method='auto'):
+        statistic_location: float
+            Value of the stochastic variable closest to the KS statistic. By
+            definition, the distance between the two sampled distributions at
+            this value is largest.
```
yes

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7617,11 +7628,23 @@ def ks_1samp(x, cdf, args=(), alternative='two-sided', method='auto'):
+        pvalue : float
+            One-tailed or two-tailed p-value.
+        statistic_location: float
+            Value of `x` corresponding with the KS statistic; i.e., the 
+            distance between the hypothesized cumulative distribution function
+            and the empirical distribution function is greatest at this
+            location.
```
Is this what you wanted `statistic_location` for `ks_1samp` to be?

It's not that. Instead, it is the CDF at which the statistic is measured.

iosonofabio replied:
Oops, that explains why the ks_2samp tests are fine but I couldn't get the one sample tests to work... Will fix tonight, thank you

mdhaber commented:
Please see iosonofabio/scipy#1.

iosonofabio commented:
Hey @mdhaber - thank you again for your PR. I merged it and added similar tests for the one-sample KS test.

I think this is ready to go in if you guys approve. Also happy to make other changes or, if you prefer, feel free to push commits into this PR directly :smile:

iosonofabio commented:
Sorry Matt, was taken in by some stuff! Let me check test results properly and I'll ask for review at that point. A day or two.

Thank you 

On Sun, Oct 2, 2022, at 09:32, Matt Haberland wrote:
> I saw you posted some new commits - did you want this to be reviewed again?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/scipy/scipy/pull/17062#issuecomment-1264497407>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJFEAHHXXCRNU6F7M7XSMLWBC3XLANCNFSM6AAAAAAQRQZWHQ>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -8177,14 +8243,30 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', method='auto'):
+        statistic_location: float
+            Value of `x` corresponding with the KS statistic; i.e. the
+            observation at which the difference between CDFs is measured.
+        statistic_sign: int
+            1 if the KS statistic is the maximal positive difference between
+            CDFs (D+), -1 if the KS statistic is the negative difference (D-).
+
```
```suggestion
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -8177,14 +8243,30 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', method='auto'):
+        statistic_location: float
+            Value of `x` corresponding with the KS statistic; i.e. the
+            observation at which the difference between CDFs is measured.
+        statistic_sign: int
+            1 if the KS statistic is the maximal positive difference between
+            CDFs (D+), -1 if the KS statistic is the negative difference (D-).
+
+        statistic_sign: int
+            1 if the KS statistic is the maximal positive difference between
+            empirical and hypothesized distributions (D+), -1 if the KS
+            statistic is the negative difference (D-). If a 2-sample test is
+            run, the difference is between the first and second empirical
+            distributions.
```
Please pick one.
If the latter, consider that the second sentence begins with "if a 2-sample test is run", but there is no corresponding part of the first sentence.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3958,13 +3975,29 @@ def test_ks1samp_allpaths(self):
         ])
         FuncData(kolmogn, dataset, (0, 1, 2), 3).check(dtypes=[int, float, bool])
 
+    @pytest.mark.parametrize("alternative, x67val, ref_location, ref_sign",
+                             [('greater', 0.5, 0.5, +1),
+                              ('less', 0.7, 0.7, -1),
+                              ('two-sided', 0.5, 0.5, +1),
+                              ('two-sided', 0.7, 0.7, -1)])
+    def test_location_sign(self, alternative, x67val, ref_location, ref_sign):
+        # Test that location and sign corresponding with statistic are as
+        # expected. (Test is designed to be easy to predict.)
+        x = np.arange(1.0, 11, dtype=np.float64) / 10.0
+        x[5:7] = x67val
```
```suggestion
    @pytest.mark.parametrize("alternative, x56val, ref_location, ref_sign",
                             [('greater', 0.5, 0.5, +1),
                              ('less', 0.7, 0.7, -1),
                              ('two-sided', 0.5, 0.5, +1),
                              ('two-sided', 0.7, 0.7, -1)])
    def test_location_sign(self, alternative, x56val, ref_location, ref_sign):
        # Test that location and sign corresponding with statistic are as
        # expected. (Test is designed to be easy to predict.)
        x = np.arange(1.0, 11, dtype=np.float64) / 10.0
        x[5:7] = x56val
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3958,13 +3975,29 @@ def test_ks1samp_allpaths(self):
+        res = stats.ks_1samp(x, stats.uniform.cdf, alternative=alternative)
+        res_list = [res.statistic, res.statistic_location, res.statistic_sign]
+        expected_list = [0.2, ref_location, ref_sign]
+        assert_array_almost_equal(np.array(res_list), np.array(expected_list))
```
```suggestion
        assert_array_almost_equal(res_list, expected_list)
```
This is not needed here or elsewhere.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3958,13 +3975,29 @@ def test_ks1samp_allpaths(self):
 
-    def _testOne(self, x1, x2, alternative, expected_statistic, expected_prob, mode='auto'):
+    def _testOne(self, x1, x2, alternative, expected_statistic, expected_prob,
+                 mode='auto', attrs=None):
```
```suggestion
                 mode='auto'):
```
?

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3890,16 +3890,33 @@ def test_agree_with_ks_1samp(self):
         self._test_kstest_and_ks1samp(x, 'greater', mode='exact')
         self._test_kstest_and_ks1samp(x, 'less', mode='exact')
 
+    def test_unpack_tuple_bunch(self):
+        x = np.linspace(-1, 1, 9)
+        res = stats.ks_1samp(x, stats.norm.cdf, alternative='two-sided')
+        ks_stat, pval = res
+        assert hasattr(res, 'statistic_location')
+        assert hasattr(res, 'statistic_sign')
+
```
Is this needed?

iosonofabio commented:
Thanks @mdhaber - implemented all requested changes. No further edits necessary AFAICT.

mdhaber reviewed (commented):
It looks like the implementation in `ks_1samp` has been corrected and documentation is better. The main comments are about the tests. It seems that what needs to be tested is essentially the same for `kstest`, `ks_1samp`, and `ks_2samp`, but what is actually tested differs. Try to simplify the changes. We should see some symmetry in the testing (whetever is done for `ks_1samp` has a parallel for `ks_2ssamp`, and somehow we see that `kstest` is the same as those two).

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -4157,8 +4173,10 @@ def testLargeBoth(self):
     def testNamedAttributes(self):
         # test for namedtuple attribute results
         attributes = ('statistic', 'pvalue')
+        additional_attributes = ('statistic_location', 'statistic_sign')
```
But why add this? It's shown that these attributes exist in both `test_location_sign` and `test_nametuples_agree`.

iosonofabio replied:
You're right, of course. It's just that I have to watch my baby while coding, which makes these things hard to catch 😅

If you were as kind, it would be just fantastic if you could push this change to the PR branch and merge so we are fine with it. Otherwise, I'll try to find time next Wednesday to come back to this... Thank you!

iosonofabio commented:
@mdhaber Thanks for the comment. I disagree that symmetry is essential. The 1-sample KS test is against a continuous distribution, whereas the 2-sample test is against a discretization thereof, which incurs all kinds of aliasing effects. I have done my best to try and satisfy the conflicting requirements of simplicity and theoretical clarity. As of now, I unfortunately have no time left for this PR. I'm not trying to be rude, but I really have no more time for this.

Given we are at the finish line, I would be grateful if you could push any changes that you think would improve the PR, symmetry or otherwise, and merge when you think it's good enough.

Thank you

mdhaber commented:
The comment about some symmetry was resolved by the two recent commits. (It really crossed paths with the first of them, even though it shows up afterwards.)

mdhaber commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -1874,13 +1874,6 @@ def test_kstest_2samp(self):
                         res3 = stats.kstest(xm, y, alternative=alternative, mode=mode)
                         assert_equal(np.asarray(res1), np.asarray(res3))
 
-    def test_nametuples_agree(self):
```
Rather than modifying this test, we can just remove it. It doesn't belong in `test_mstats_basic.py` if it's about `stats` objects anyway, and `Ks_2sampResult` is no longer used either in `stats` or `mstats`, so I think it's safe to remove this.

mdhaber reviewed (approved):
Approved, but I'd like a second opinion before merging, since I had a lot to do with the documentation and tests.

tirthasheshpatel reviewed (approved):
Looks good to me too! Thanks @iosonofabio @mdhaber! I initially thought that a new function would be better but, looking at this PR, it seems like either work just fine (with respect to API design choice). Since I haven't gone through all the reviews and comments on this PR, I would leave the decision to merge on @mdhaber.

mdhaber commented:
Thanks @tirthasheshpatel! @tupui I'm going to merge now but if you have a chance for a quick glance, let me know if there is anything you'd like to change before the release and I'll open another PR.

iosonofabio commented:
Thanks @mdhaber and everyone for your hard work and guidance!
